File: java/2559.java
class Solution {
    private boolean vowel(char c) {
        return (c == 'a') || (c == 'e') || (c == 'i') || (c == 'o') || (c == 'u');
    }

    private boolean vs(String s) {
        return vowel(s.charAt(0)) && vowel(s.charAt(s.length() - 1));
    }

    public int[] vowelStrings(String[] w, int[][] q) {
        int n = w.length;
        int psa[] = new int[n];
        int ps = 0;
        for (int i = 0; i < n; i++) {
            if (vs(w[i]))
                ++ps;
            psa[i] = ps;
        }
        n = q.length;
        int res[] = new int[n];
        for (int i = 0; i < n; i++)
            if (q[i][0] == 0)
                res[i] = psa[q[i][1]];
            else
                res[i] = psa[q[i][1]] - psa[q[i][0] - 1];
        return res;
    }
}

File: javascript/2559.js
/**
 * @param {string[]} w
 * @param {number[][]} q
 * @return {number[]}
 */
var vowelStrings = function (w, q) {
    let v = new Set(['a', 'e', 'i', 'o', 'u'])
    let psa = [0]
    for (let i = 0; i < w.length; i++)
        psa.push(psa.at(i) + ((v.has(w[i][0]) && v.has(w[i][w[i].length - 1])) ? 1 : 0))
    let res = []
    for (let j of q)
        res.push(psa[j[1] + 1] - psa[j[0]])
    return res
};
